### ISSUE TITLE
Random head-sampling broken

### ISSUE THREAD

**Issue by Sibirius** [maraisr/rian#5]
### Describe the bug

We have implemented a simple random head sampler:
```ts
export class Sampler {
  constructor(private readonly sampleRate: number) {}

  public sample: Rian.Sampler = (_, parentId?: tctx.Traceparent) => {
    if (this.sampleRate === 0) {
      return false;
    }

    if (!parentId) {
      return Math.random() <= this.sampleRate;
    }

    return tctx.is_sampled(parentId);
  };
}
```

This doesn't work anymore due to breaking changes in `rian` and `tctx`.

1. `rian` changed the sampler to get the new span id instead of its parent id:
https://github.com/maraisr/rian/commit/724248d4de759635a20fb7a523117f7eaef4e2fc

This makes it impossible to differentiate in the sampler between a fresh "root" span that was created using `tctx.make()` and being in a child span. That means `parentId` will always be set and we can't know whether we should random sample or not.

We are able to work around this by doing the random sampling outside of the sampler and passing our pre-sampled traceparent into the `scope()` call to avoid the ` || tctx.make()` branch. This feels a bit hacky though and also doesn't actually work due to the next issue.

2. `tctx` changed the behavior when creating a child span: from inheriting flags, to disabling the sampled flag and then to always enable it:
https://github.com/maraisr/tctx/commit/fe96b67a49599b3dda10427edfbcee4ca73d86ff
https://github.com/maraisr/tctx/commit/05d0acf106689e50d2b41d9ce50a9280ee1ff69d

This means that the child which is passed to the sampler always has the sampled flag activated. https://github.com/maraisr/rian/blob/main/src/index.ts#L25

So the sampler has no way of inheriting the sampling status of the parent: it doesn't have access to the parent and the child lost all information about the sampled flag.

Our temporary solution is a local patch to pass the `parent` into the sampler instead of the `id`.

But since the change from `parent` to `id` was a deliberate commit, I didn't want to just open a PR. I'm probably missing some context for that decision.
The same goes for `tctx` as that also seemed to be a deliberate decision to not inherit flags anymore as it's even stated in the tests.

### To reproduce

Try sampling using either the sampler I provided or even the `defaultSampler`.
When passing in a traceparent which ends in `-00` (not sampled) it will still produce a span which is sampled.

### Expected behavior

A newly created span should be able to inherit the sampled flag from its traceparent.

**Comment by maraisr** [comment]
@Sibirius, thank you for raising this issue. ❤ I apologize for the delayed response.

<details><summary>tctx issues</summary>

The issue there likely stems from the introduction of the [`random`](https://github.com/maraisr/tctx/commit/fbbc5cb4ce735fb09e96c6acdd76fa714dc173e1) flag (new level 3 spec). My apologies for this oversight.

To clarify, tctx can't determine how to handle the child's sampling flag (whether the ID will be recorded, decision deferred, etc.). The best approach is to forward it from the parent, as it should have done originally. And leave that decision to our component (as in rian).

According to the spec, when mutating the sample flag, you [must also update](https://github.com/w3c/trace-context/blob/main/spec/20-http_request_header_format.md#sampled-flag) the `parent-id`. Which was why we had the `sampled` on the child method, but to avoid endless arguments on that method (now with random as well) each with different tradeoffs, I moved it to a util, but incorrectly force-defaulted it. Whoops!

</details>

When creating a new span, the span's ID sampled flag is irrelevant until we get an answer from the sampler. So we should just copy from the parent. Fixed in https://github.com/maraisr/tctx/commit/8ec6b3ee7d7ce3ef557e5145cbfd4746a28edf0a.

Previously, we provided only the parent ID to the sampler, which had returned the sampling state for the new span before creation. But that did not allow any sampling decision to be made with respect to current/new span id (which was actually influenced by how we sample our spans at work, sample spans, than entire traces). So I moved to the new id, only not really thinking broader enough. Sorry!

As you've pointed out, there are cases where sampling decisions are based off the parent. Combined with not to force sample its ID. We are the "component" here in rian making the decisions, and should not have been tctx.

I am considering changing the rian sampler to:

```ts
function sampler(id: string, parent: Traceparent|undefined, name: string, tracer: TracerScope): boolean
```

This will allow you to:

- Inspect the parent (or lack thereof) to make a decision
- Use a random `Math.random()`-style decision
- Apply a [`murmur hash`](https://en.wikipedia.org/wiki/MurmurHash) (or similar) on the ID
- Based the decision on the name of the span or scope

I believe this approach would restore your sampler's functionality without requiring hacks. I've pushed that change to [`rian@0.4.0-beta.2`](https://www.npmjs.com/package/rian/v/0.4.0-beta.2), please let me know if that works for you. (change PR: https://github.com/maraisr/rian/pull/6)

**Comment by maraisr** [comment]
I have implemented the above, and merged. And is now published as `v0.4.0`.

Your sampler above, now works as expected.

<details><summary>Sample code</summary>

```ts
import { tracer, report, configure } from 'rian';
import { exporter } from 'rian/exporter.console';

import { is_sampled } from 'tctx/traceparent';
import type * as Rian from 'rian';

configure('my service');

class Sampler {
  constructor(private readonly sampleRate: number) {}

  public sample: Rian.Sampler = (_, parentId) => {
    if (this.sampleRate === 0) {
      return false;
    }

    if (!parentId) {
      return Math.random() <= this.sampleRate;
    }

    return is_sampled(parentId);
  };
}

let sampler = new Sampler(0.5);

let t = tracer('api', {
  sampler: sampler.sample.bind(sampler),
});

for (let i = 0; i < 20; i++) {
  t.span('req')((s) => {
    s.span('db')(() => {});
  });
}

report(exporter(process.stdout.columns));
```

</details>

**Comment by Sibirius** [comment]
Hey @maraisr, thank you so much for the fix and the detailed response <3

This time I'm sorry for the delayed response. I tried to squeeze this in over the last week but only finally managed to find the time yesterday. Unfortunately I then ran into some type issues with types from `tctx` being treated as `any`, so I couldn't properly test it yet.
I'll open a separate issue for the types, once I figure out a bit better what's happening.
